connect-inject/health_check_resource_test.go

```diff
@@ -0,0 +1,606 @@
+	"github.com/hashicorp/consul/sdk/testutil/retry"
+	"github.com/hashicorp/go-hclog"
+	"github.com/stretchr/testify/require"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes/fake"
+)
+
+const (
+	testPodName               = "test-pod"
+	testServiceNameAnnotation = "test-service"
+	testServiceNameReg        = "test-pod-test-service"
+	testHealthCheckID         = "default_test-pod-test-service_kubernetes-health-check-ttl"
+	testFailureMessage        = "Kubernetes pod readiness probe failed"
+	testDoNotRegister         = "do not register"
+
+	testCheckNotesPassing  = "Kubernetes Health Checks Passing"
+	testTypesBoth          = "both"
+	testTypesUpsertOnly    = "upsert"
+	testTypesReconcileOnly = "reconcile"
+	testUpsert             = "upsert"
+	testReconcile          = "reconcile"
+)
+
+func getSupportedTestTypes(testTypes string) map[string]bool {
+	switch testTypes {
+	case testTypesBoth:
+		return map[string]bool{testUpsert: true, testReconcile: true}
+	case testTypesReconcileOnly:
+		return map[string]bool{testReconcile: true}
+	case testTypesUpsertOnly:
+		return map[string]bool{testUpsert: true}
+	}
+	return nil
+}
+
+var testPodSpec = corev1.PodSpec{
+	Containers: []corev1.Container{
+		corev1.Container{
+			Name: testPodName,
+		},
+	},
+}
+
+func testServerAgentResourceAndController(t *testing.T, pod *corev1.Pod) (*testutil.TestServer, *api.Client, *HealthCheckResource) {
+	require := require.New(t)
+	// Set up server, client
+	s, err := testutil.NewTestServerConfigT(t, nil)
+	require.NoError(err)
+
+	clientConfig := &api.Config{Address: s.HTTPAddr}
+	require.NoError(err)
+	client, err := api.NewClient(clientConfig)
+	require.NoError(err)
+
+	schema := "http://"
+	consulUrl, err := url.Parse(schema + s.HTTPAddr)
+	require.NoError(err)
+
+	healthResource := HealthCheckResource{
+		Log:                 hclog.Default().Named("healthCheckResource"),
+		KubernetesClientset: fake.NewSimpleClientset(pod),
+		ConsulUrl:           consulUrl,
+		ReconcilePeriod:     0,
+	}
+	return s, client, &healthResource
+}
+
+func registerHealthCheck(t *testing.T, client *api.Client, initialState, reason string) {
+	require := require.New(t)
+	err := client.Agent().CheckRegister(&api.AgentCheckRegistration{
+		Name:      "Kubernetes Health Check",
+		ID:        testHealthCheckID,
+		ServiceID: testServiceNameReg,
+		Notes:     reason,
+		AgentServiceCheck: api.AgentServiceCheck{
+			TTL:    "100000h",
+			Status: initialState,
+			Notes:  reason,
+		},
+	})
+	require.NoError(err)
+}
+
+// We expect to already be pointed at the correct agent
+func getConsulAgentChecks(t *testing.T, client *api.Client) *api.AgentCheck {
+	require := require.New(t)
+	filter := fmt.Sprintf("CheckID == `%s`", testHealthCheckID)
+	checks, err := client.Agent().ChecksWithFilter(filter)
+	require.NoError(err)
+	return checks[testHealthCheckID]
+}
+
+func TestHealthCheckHandlers(t *testing.T) {
```

Comment:
Can we add the keys explicitly to the test cases please? It was hard for me to look at a case and understand what the implication of certain bool values were. This probably won't be a problem in an IDE I guess, but would be nice. This comment is not a blocker.